### PR TITLE
boards: arm: add stm32f103_mini board 

### PR DESCRIPTION
**Pull request by wuhanstudio**
Add a popular STM32F103 mini board that has multiple variations:

stm32f103_mini_yellow

stm32f103_mini_blue

Tested LED and UART related examples:

t4fx6weu

### CODE REVIEW

File: boards/arm/stm32f103_mini/Kconfig.board [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# STM32F103RCT6 Mini board configuration
+
+# Copyright (c) 2020 WuhanStudio
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_STM32F103_MINI
+	bool "STM32F103RCT6 Mini Board"
+	depends on SOC_STM32F103XB
```
Comment (erwango) [inline boards/arm/stm32f103_mini/Kconfig.board:8]: According to HAL, that should actually be SOC_STM32F103XE (cf https://github.com/zephyrproject-rtos/hal_stm32/blob/master/stm32cube/stm32f1xx/soc/stm32f1xx.h#L73)

File: boards/arm/stm32f103_mini/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+- USB re-enumeration capability:
+
+  - Mass storage
+
+- Support of wide choice of Integrated Development Environments (IDEs) including:
+
+  - IAR
+  - ARM Keil
+  - GCC-based IDEs
```
Comment (erwango) [inline boards/arm/stm32f103_mini/doc/index.rst:48]: This part is not pertinent for a Zephyr documentation, please remove.

File: boards/arm/stm32f103_mini/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+flashed in the usual way (see :ref:`build_an_application` and
+:ref:`application_run` for more details).
+
+Flashing
+========
+
+There are 2 main entry points for flashing STM32F1X SoCs, one using the ROM
+bootloader, and another by using the SWD debug port (which requires additional
+hardware). Flashing using the ROM bootloader requires a special activation
+pattern, which can be triggered by using the BOOT0 pin.
```
Comment (erwango) [inline boards/arm/stm32f103_mini/doc/index.rst:127]: In (boards/arm/stm32f103_mini/board.cmake)[https://github.com/zephyrproject-rtos/zephyr/pull/30177/commits/13dceada50ac2df204c90a96dba881332acca677#diff-f4cb4e26b5274d7bc7b78b8a6d0d651c330ccfe0ef4585422f9f79cba7736376], you configured openocd/jlink, which is differet from what you're describing here. Can you clarify how this is supposed to work?

File: boards/arm/stm32f103_mini/stm32f103_mini.dts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+&spi2 {
+	pinctrl-0 = <&spi2_nss_master_pb12 &spi2_sck_master_pb13
+		     &spi2_miso_master_pb14 &spi2_mosi_master_pb15>;
+	status = "okay";
+};
+
+&iwdg {
+		status = "okay";
```
Comment (erwango) [inline boards/arm/stm32f103_mini/stm32f103_mini.dts:70]: Please fix indentation

File: boards/arm/stm32f103_mini/stm32f103_mini.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  - gnuarmemb
+  - xtools
+ram: 48
+flash: 256
+supported:
+  - gpio
+  - spi
+  - pwm
+  - watchdog
+  - adc
```
Comment (erwango) [inline boards/arm/stm32f103_mini/stm32f103_mini.yaml:16]: i2c is missing

File: dts/arm/st/f1/stm32f103Xc.dtsi [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+			status = "disabled";
+			label = "CAN_1";
+		};
+	};
+
+	usb_fs_phy: usbphy {
+		compatible = "usb-nop-xceiv";
+		#phy-cells = <0>;
+		label = "USB_FS_PHY";
+	};
```
Comment (erwango) [inline dts/arm/st/f1/stm32f103Xc.dtsi:70]: This part is a bit tricky:
Looking at peripherals, a STM32F103xC is similar to STM32F103xE: a STM32X103xB + Tim(5,6,7,8) + ADC3 + SPI3 + DMA2 + UART(4,5).
Difference bewteen STM32F103xE and STM32F103xC is the flash/RAM:
- STM32F103xC: 48 RAM, 256 Flash
- STM32F103xE: 64 RAM, 512 Flash

So, if you're introducing STM32F103xC, here is how to proceed:
- Rename current stm32f103xe.dtsi to stm32f103xc.dtsi, and change RAM flash sizes to 48 RAM, 256 Flash
- Add stm32f103xe.dtsi file which includes stm32f103xc.dtsi and overwrites  RAM flash sizes to 64 RAM, 512 Flash

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this addition!
Adding a new stm32f103 soc is a bit tricky, so please follow may advices.
Also please clarify the flash procedure.

File: boards/arm/stm32f103_mini/doc/index.rst (hunk 3 above)
Comment (wuhanstudio) [reply boards/arm/stm32f103_mini/doc/index.rst:127]: > In (boards/arm/stm32f103_mini/board.cmake)[https://github.com/[/pull/30177](https://github.com/zephyrproject-rtos/zephyr/pull/30177)/commits/13dceada50ac2df204c90a96dba881332acca677#diff-f4cb4e26b5274d7bc7b78b8a6d0d651c330ccfe0ef4585422f9f79cba7736376], you configured openocd/jlink, which is differet from what you're describing here. Can you clarify how this is supposed to work?

Hi, in **board.cmake**, I configured openocd / st-link which uses the SWD debug port as described in **index.rts**.

And the other method uses serial port (RX/TX) which requires tools from ST as described here:

[http://manpages.ubuntu.com/manpages/cosmic/man1/st-flash.1.html](http://manpages.ubuntu.com/manpages/cosmic/man1/st-flash.1.html)

Is it OK to include only st-link in board which is a much better choice, because st-flash does not support Debug?

File: dts/arm/st/f1/stm32f103Xc.dtsi (hunk 6 above)
Comment (wuhanstudio) [reply dts/arm/st/f1/stm32f103Xc.dtsi:70]: Indeed, this is a better way, thank you.

I'm working on this.

File: boards/arm/stm32f103_mini/doc/index.rst (hunk 3 above)
Comment (erwango) [reply boards/arm/stm32f103_mini/doc/index.rst:121]: Sure, ST-link is fine, but please mention it in index.rst

File: boards/arm/stm32f103_mini/Kconfig.board (hunk 1 above)
Comment (wuhanstudio) [reply boards/arm/stm32f103_mini/Kconfig.board]: Sorry, changed it.

File: boards/arm/stm32f103_mini/doc/index.rst (hunk 2 above)
Comment (wuhanstudio) [reply boards/arm/stm32f103_mini/doc/index.rst]: Just removed it.

File: boards/arm/stm32f103_mini/stm32f103_mini.dts (hunk 4 above)
Comment (wuhanstudio) [reply boards/arm/stm32f103_mini/stm32f103_mini.dts]: Sorry, did not notice, and it's fixed now.

File: boards/arm/stm32f103_mini/stm32f103_mini.yaml (hunk 5 above)
Comment (wuhanstudio) [reply boards/arm/stm32f103_mini/stm32f103_mini.yaml]: Forgot i2c, and it's there now.

**Comment by wuhanstudio** [comment]
> Thanks for this addition!
> Adding a new stm32f103 soc is a bit tricky, so please follow may advices.
> Also please clarify the flash procedure.

Thank you for your review and pointing out those issues, and I have fixed them. Looking forward to your furthur opinions, thanks again.

File: tests/drivers/adc/adc_api/src/test_adc.c [hunk 7]
```diff
@@ -155,6 +155,7 @@
 	defined(CONFIG_BOARD_NUCLEO_L152RE) || \
 	defined(CONFIG_BOARD_OLIMEX_STM32_H103) || \
 	defined(CONFIG_BOARD_96B_AEROCORE2) || \
+	defined(CONFIG_BOARD_STM32F103_MINI) || \
```
Comment (erwango) [inline tests/drivers/adc/adc_api/src/test_adc.c:158]: Please move this change in a dedicated commit

File: tests/kernel/gen_isr_table/testcase.yaml [hunk 8]
```diff
@@ -4,7 +4,7 @@ tests:
     filter: CONFIG_GEN_ISR_TABLES and CONFIG_ARMV6_M_ARMV8_M_BASELINE
     tags: interrupt isr_table
   arch.interrupt.gen_isr_table.arm_mainline:
-    platform_exclude: nucleo_f103rb olimexino_stm32 stm32_min_dev_black
+    platform_exclude: stm32f103_mini nucleo_f103rb olimexino_stm32 stm32_min_dev_black
```
Comment (erwango) [inline tests/kernel/gen_isr_table/testcase.yaml:7]: Please move this change in a dedicated commit

**Review by erwango — CHANGES_REQUESTED** [review]
Seems fine now.
Please changes to the samples and tests in dedicated commits, but otherwise, this is good!

File: tests/drivers/adc/adc_api/src/test_adc.c (hunk 7 above)
Comment (wuhanstudio) [reply tests/drivers/adc/adc_api/src/test_adc.c:158]: No problem.

File: tests/kernel/gen_isr_table/testcase.yaml (hunk 8 above)
Comment (wuhanstudio) [reply tests/kernel/gen_isr_table/testcase.yaml]: OK

**Review by erwango — APPROVED** [review]

File: dts/arm/st/f1/stm32f103Xc.dtsi [hunk 9]
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2017 I-SENSE group of ICCS
+ *
+ * SoC device tree include for STM32F103xE SoCs
```
Comment (ABOSTM) [inline dts/arm/st/f1/stm32f103Xc.dtsi:4]: STM32F103x**C**

File: boards/arm/stm32f103_mini/doc/index.rst [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+| GPIO      | on-chip    | gpio                                |
++-----------+------------+-------------------------------------+
+| CLOCK     | on-chip    | reset and clock control             |
++-----------+------------+-------------------------------------+
+| FLASH     | on-chip    | flash memory                        |
++-----------+------------+-------------------------------------+
+| WATCHDOG  | on-chip    | independent watchdog                |
++-----------+------------+-------------------------------------+
+| ADC       | on-chip    | ADC Controller                      |
++-----------+------------+-------------------------------------+
```
Comment (ABOSTM) [inline boards/arm/stm32f103_mini/doc/index.rst:73]: Missing PWM and SPI

File: boards/arm/stm32f103_mini/doc/index.rst [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+   :height: 619px
+   :alt: Nucleo F103RB connectors
+
+Default Zephyr Peripheral Mapping:
+----------------------------------
+
+- UART_1 TX/RX : PA9/PA10
+- UART_2 TX/RX : PA2/PA3 (ST-Link Virtual COM Port)
+- SPI1 NSS/SCK/MISO/MOSI : PA4/PA5/PA6/PA7 (Arduino SPI)
+- SPI2 SCK/MISO/MOSI : PB12/PB13/PB14/PB15
```
Comment (ABOSTM) [inline boards/arm/stm32f103_mini/doc/index.rst:102]: **NSS**/SCK/MISO/MOSI

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: dts/arm/st/f1/stm32f103Xc.dtsi (hunk 9 above)
Comment (wuhanstudio) [reply dts/arm/st/f1/stm32f103Xc.dtsi:4]: stm32f103Xc.dtsi is based on previous stm32f103Xe.dtsi. 
A real mistake. Working on it.

File: boards/arm/stm32f103_mini/doc/index.rst (hunk 10 above)
Comment (wuhanstudio) [reply boards/arm/stm32f103_mini/doc/index.rst:73]: Thanks, it has been added now.

File: boards/arm/stm32f103_mini/doc/index.rst (hunk 11 above)
Comment (wuhanstudio) [reply boards/arm/stm32f103_mini/doc/index.rst]: OK

**Review by ABOSTM — APPROVED** [review]
LGTM

**Review by MaureenHelm — APPROVED** [review]
